File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.cpp
#include <string>
#include <vector>
#include <map>
#include <sstream>

using namespace std;

vector<string> solution(vector<string> record) {
    vector<pair<string, string>> log;
    vector<string> answer;
    map<string, string> m;
    string command, uid, name;
    for (int i = 0; i < record.size(); i++) {
        stringstream ss(record[i]);
        ss >> command >> uid >> name;
    
        if (command == "Enter" || command == "Change")
            m[uid] = name;
        if (command == "Enter" || command == "Leave")
            log.emplace_back(uid, command);
    }

    for (int i = 0; i < log.size(); i++) {
        if (log[i].second == "Enter")
            answer.push_back(m[log[i].first] + "님이 들어왔습니다.");
        else
            answer.push_back(m[log[i].first] + "님이 나갔습니다.");
    }

    return answer;
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    
    static Map<String, String> map = new HashMap<>();
    
    public String[] solution(String[] records) {  
        int cnt = 0;
        
        for (String record : records) {
            String[] split = record.split(" ");
            
            String cmd = split[0];
            
            if (cmd.equals("Enter")) {
                map.put(split[1], split[2]);
                cnt++;
            } else if (cmd.equals("Change")) {
                map.put(split[1], split[2]);
            } else if (cmd.equals("Leave")) {
                cnt++;
            }
        }
        
        String[] answer = new String[cnt];   
        
        int idx = 0;
        for (String record : records) {
            String[] split = record.split(" ");

            String cmd = split[0];
            
            if (cmd.equals("Enter")) {
                answer[idx++] = String.format("%s님이 들어왔습니다.", map.get(split[1]));
            } else if (cmd.equals("Leave")) {
                answer[idx++] = String.format("%s님이 나갔습니다.", map.get(split[1]));
            }
        }
        
        return answer;
    }
}